modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.h

```diff
@@ -0,0 +1,299 @@
+  unsigned int GetCols() const { return cols_; }
+
+  /**@brief Combine two NdtMapMatrix instances (Reduce). */
+  static void Reduce(NdtMapMatrix* cells, const NdtMapMatrix& cells_new);
+
+ private:
+  /**@brief The number of rows. */
+  unsigned int rows_;
+  /**@brief The number of columns. */
+  unsigned int cols_;
+  /**@brief The matrix data structure. */
+  NdtMapCells* map3d_cells_;
+};
+
+inline void NdtMapSingleCell::Reset() {
+  intensity_ = 0.0;
+  intensity_var_ = 0.0;
+  road_pt_count_ = 0;
+  count_ = 0;
+  centroid_ = Eigen::Vector3f::Zero();
+  centroid_average_cov_ = Eigen::Matrix3f::Zero();
+  centroid_icov_ = Eigen::Matrix3f::Identity();
+  is_icov_available_ = false;
+}
+
+inline void NdtMapSingleCell::AddSample(const float intensity,
+                                        const float altitude,
+                                        const Eigen::Vector3f& centroid,
+                                        bool is_road) {
+  ++count_;
+  float v1 = intensity - intensity_;
+  intensity_ += v1 / static_cast<float>(count_);
+  float v2 = intensity - intensity_;
+  intensity_var_ = (static_cast<float>(count_ - 1) * intensity_var_ + v1 * v2) /
+                   static_cast<float>(count_);
+
+  if (is_road) {
+    ++road_pt_count_;
+  }
+
+  Eigen::Vector3f v3 = centroid - centroid_;
+  centroid_ += v3 / static_cast<float>(count_);
+  Eigen::Matrix3f v4 = centroid * centroid.transpose() - centroid_average_cov_;
+  centroid_average_cov_ += v4 / static_cast<float>(count_);
+  CentroidEigenSolver(centroid_average_cov_);
+}
+
+inline void NdtMapSingleCell::MergeCell(const float intensity,
+                                        const float intensity_var,
+                                        const unsigned int road_pt_count,
+                                        const unsigned int count,
+                                        const Eigen::Vector3f& centroid,
+                                        const Eigen::Matrix3f& centroid_cov) {
+  unsigned int new_count = count_ + count;
+  float p0 = static_cast<float>(count_) / static_cast<float>(new_count);
+  float p1 = static_cast<float>(count) / static_cast<float>(new_count);
+  float intensity_diff = intensity_ - intensity;
+
+  intensity_ = intensity_ * p0 + intensity * p1;
+  intensity_var_ = intensity_var_ * p0 + intensity_var * p1 +
+                   intensity_diff * intensity_diff * p0 * p1;
+
+  centroid_[0] = centroid_[0] * p0 + centroid[0] * p1;
+  centroid_[1] = centroid_[1] * p0 + centroid[1] * p1;
+  centroid_[2] = centroid_[2] * p0 + centroid[2] * p1;
+
+  count_ = new_count;
+  road_pt_count_ += road_pt_count;
+}
+
+inline void NdtMapSingleCell::MergeCell(const NdtMapSingleCell& cell_new) {
+  MergeCell(cell_new.intensity_, cell_new.intensity_var_,
+            cell_new.road_pt_count_, cell_new.count_, cell_new.centroid_,
+            cell_new.centroid_average_cov_);
+}
+
+inline void NdtMapSingleCell::CentroidEigenSolver(
```

Comment:
Avoid inline functions with complicated logic. This applies to all where applicable.